pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -0,0 +1,182 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+
+	drapbv1 "k8s.io/kubelet/pkg/apis/dra/v1alpha1"
+)
+
+type Client interface {
+	NodePrepareResource(
+		ctx context.Context,
+		namespace string,
+		claimUID types.UID,
+		claimName string,
+		resourceHandle string,
+	) (*drapbv1.NodePrepareResourceResponse, error)
+
+	NodeUnprepareResource(
+		ctx context.Context,
+		namespace string,
+		claimUID types.UID,
+		claimName string,
+		cdiDevice []string,
+	) (*drapbv1.NodeUnprepareResourceResponse, error)
+}
+
+// Strongly typed address.
+type draAddr string
+
+// draPluginClient encapsulates all dra plugin methods.
+type draPluginClient struct {
+	pluginName          string
+	addr                draAddr
+	nodeV1ClientCreator nodeV1ClientCreator
+}
+
+var _ Client = &draPluginClient{}
+
+type nodeV1ClientCreator func(addr draAddr) (
+	nodeClient drapbv1.NodeClient,
+	closer io.Closer,
+	err error,
+)
+
+// newV1NodeClient creates a new NodeClient with the internally used gRPC
+// connection set up. It also returns a closer which must be called to close
+// the gRPC connection when the NodeClient is not used anymore.
+// This is the default implementation for the nodeV1ClientCreator, used in
+// newDRAPluginClient.
+func newV1NodeClient(addr draAddr) (nodeClient drapbv1.NodeClient, closer io.Closer, err error) {
+	var conn *grpc.ClientConn
+
+	conn, err = newGrpcConn(addr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return drapbv1.NewNodeClient(conn), conn, nil
+}
+
+func NewDRAPluginClient(pluginName string) (Client, error) {
+	if pluginName == "" {
+		return nil, fmt.Errorf("plugin name is empty")
+	}
+
+	existingPlugin := draPlugins.Get(pluginName)
+	if existingPlugin == nil {
+		return nil, fmt.Errorf("plugin name %s not found in the list of registered DRA plugins", pluginName)
+	}
+
+	return &draPluginClient{
+		pluginName:          pluginName,
+		addr:                draAddr(existingPlugin.endpoint),
+		nodeV1ClientCreator: newV1NodeClient,
+	}, nil
+}
+
+func (r *draPluginClient) NodePrepareResource(
+	ctx context.Context,
+	namespace string,
+	claimUID types.UID,
+	claimName string,
+	resourceHandle string,
+) (*drapbv1.NodePrepareResourceResponse, error) {
+	klog.V(4).InfoS(
+		log("calling NodePrepareResource rpc"),
+		"namespace", namespace,
+		"claim UID", claimUID,
+		"claim name", claimName,
+		"resource handle", resourceHandle)
+
+	if r.nodeV1ClientCreator == nil {
+		return nil, errors.New("failed to call NodePrepareResource. nodeV1ClientCreator is nil")
+	}
+
+	nodeClient, closer, err := r.nodeV1ClientCreator(r.addr)
```

Comment:
It seems like on every call to `NodePrepareResource` and `NodeUnprepareResource` we create a new GRPC connection? Does it make sense to have a single connection that we keep open and re-use? Or is this deliberately done this way?

Reply:
As the code is currently written, caching this gRPC connection in the `draPluginClient` would not provide any benefit because a new `draPluginClient` is created *every* time a `PrepareResource()` call is made in the `draManager`. 

During my review, I suggested caching `draPluginClient`s in the `draManager` (instead of reecrating them for each call to `PrepareResource()`).
https://github.com/kubernetes/kubernetes/pull/111023#discussion_r1009408979

Only if this is done, would it make sense to cache the gRPC client as you are suggesting here as well.

I believe it's a worthwhile change, but I don't feel too strongly about it for the `alpha` release. If you feel strongly, then I can work with @bart0sh to make the change.

Reply:
created an issue in the DRA issue tracker: https://github.com/orgs/kubernetes/projects/95/views/1 so it's not forgotten.

Reply:
Sounds good, agree this is not blocking, but thanks for creating issue so this is considered for beta.